### PR TITLE
KIALI-1423 Move Jaeger iFrame to external links

### PR DESCRIPTION
**Pull request by aljesusg**
** Describe the change **

Remove proxy configuration for Jaeger

UI Side https://github.com/kiali/kiali-ui/pull/619
** Issue reference **
[KIALI-1423](
https://issues.jboss.org/browse/KIALI-1423)
### Correct Information
grafana_correct
### Grafana configuration needed
grafana_need_config
### Grafana URL bad format (Example: Need schema)
grafana_bad_format

### Jaeger correct
jaeger_correct
### Jaeger  configuration needed
jaeger_no_config
### Jaeger URL bad format (Example: Need schema)
jaeger_mal_format

## Launch with the kiali-configmap

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kiali
  labels:
    app: kiali
    version: ${VERSION_LABEL}
data:
  config.yaml: |
    server:
      port: 20001
      static_content_root_directory: /opt/kiali/console
    external_services:
      jaeger:
        url: https://jaeger-query-istio-system.127.0.0.1.nip.io
      grafana:
        url: http://grafana-istio-system.127.0.0.1.nip.io
```

### CODE REVIEW

**Comment by aljesusg** [comment]
@hhovsepy Can you verify this? Thanks

**Comment by hhovsepy** [comment]
checking

File: handlers/jaeger.go [hunk 1]
```diff
@@ -3,42 +3,44 @@ package handlers
...
 func GetJaegerInfo(w http.ResponseWriter, r *http.Request) {
 	scheme := "http"
 	if r.URL.Scheme != "" {
 		scheme = r.URL.Scheme
 	}
-	// r.Host can be set as "host" or "host:port", we always remove the :port as it is updated by the proxy
-	kialiHost := strings.Split(r.Host, ":")
+	jaegerConfig := config.Get().ExternalServices.Jaeger
+
+	spec, err := getService(jaegerConfig.ServiceNamespace, jaegerConfig.Service)
```
Comment (lucasponce) [inline handlers/jaeger.go]: I am not sure if we should fully delegate this into the configuration.
I mean, Istio by default does not expose Jaeger, so it is clear that we should ask a step to the user to publish the URL.
As community is using kubernetes on several vendors, perhaps it's just a matter to indicate this step instead to try to resolve it automatically.
I do not know, I guess this happens same as graphana links, as it is same scenario.

Comment (aljesusg) [reply handlers/jaeger.go]: We are doing the same for the grafana info. So we should change for grafana too.

**Comment by hhovsepy** [comment]
Jeager failed to load:
screenshot from 2018-08-27 20-03-17

**Comment by aljesusg** [comment]
Yes @hhovsepy the UI PR with this one is this https://github.com/kiali/kiali-ui/pull/619 , the behaviour will be a new tab with the jaeger URL like this gif

jaeeg

**Comment by hhovsepy** [comment]
kiali-ui 0.7.0 (8cf02cce484b795b1958b9ccc9526132e178c881)
kiali v0.7.0-SNAPSHOT (76028d847227b14b482b66a727c8e8a68c18c284)

JS error when clicking on Distributed Tracing menu:
screenshot from 2018-08-28 07-19-28

JS error when clicking on Traces tab:
screenshot from 2018-08-28 07-19-47

Same with Chrome:
screenshot from 2018-08-28 07-20-38

**Comment by aljesusg** [comment]
Ok @hhovsepy I talked with @lucasponce  about and we did some changes now We need to set the URl in the configuration

## Grafana
### Correct Information
grafana_correct
### Grafana configuration needed
grafana_need_config
### Grafana URL bad format (Example: Need schema)
grafana_bad_format

## Jaeger

### Jaeger correct
jaeger_correct
### Jaeger  configuration needed
jaeger_no_config
### Jaeger URL bad format (Example: Need schema)
jaeger_mal_format


## Launch with the kiali-configmap

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kiali
  labels:
    app: kiali
    version: ${VERSION_LABEL}
data:
  config.yaml: |
    server:
      port: 20001
      static_content_root_directory: /opt/kiali/console
    external_services:
      jaeger:
        url: https://jaeger-query-istio-system.127.0.0.1.nip.io
      grafana:
        url: http://grafana-istio-system.127.0.0.1.nip.io
```

**Comment by hhovsepy** [comment]
@aljesusg it worked after config change, verified.

**Comment by aljesusg** [comment]
Great Thanks @hhovsepy

File: README.adoc [hunk 2]
```diff
@@ -460,26 +460,26 @@ external_services:
     var_workload: VALUE
 ----
 
-|`JAEGER_SERVICE_NAMESPACE`
-|The Kubernetes namespace that holds the Jaeger service. (default is `istio-system`)
+|`JAEGER_URL`
+|The URL to the Jaeger server. When not set, the URL may be automatically detected from OpenShift or Kubernetes API.
```
Comment (lucasponce) [inline README.adoc]: "The URL to the Jaeger Query service." (As jaeger has several services, "server" perhaps could be ambiguous).

> When not set, the URL may be automatically detected from OpenShift or Kubernetes API.

This is not defined, if we want to replicate same logic as used for Grafana if the URL is not setup, I am fine with that, but then Grafana & Jaeger should follow same logic as basically both are similar external services.

File: README.adoc [hunk 3]
```diff
@@ -503,6 +503,73 @@ login_token:
...
+
+=== Jaeger
+
+To configure the Jaeger Service you need to set the URL parameter in the config.yaml
+
+[source,yaml]
+----
+external_services:
+  jaeger:
+    url: https://jaeger-query-istio-system.127.0.0.1.nip.io
```
Comment (lucasponce) [inline README.adoc]: "https" by default could lead that people perform copy/paste and add it.

File: config.yaml [hunk 4]
```diff
@@ -21,7 +21,10 @@ external_services:
   # Uncomment istio_identity_domain to set a different value. This value must match the Istio configuration.
   # Default is "svc.cluster.local" (which matches Istio's default)
   # istio_identity_domain: svc.cluster.local
-
+  jaeger:
+    url: https://jaeger-query-istio-system.127.0.0.1.nip.io
```
Comment (lucasponce) [inline config.yaml]: Same here, developers installation are not usually deployed under "https" scheme, I think.

File: deploy/openshift/kiali-configmap.yaml [hunk 5]
```diff
@@ -10,3 +10,8 @@ data:
     server:
       port: 20001
       static_content_root_directory: /opt/kiali/console
+    external_services:
+      jaeger:
+        url: https://jaeger-query-istio-system.127.0.0.1.nip.io
```
Comment (lucasponce) [inline deploy/openshift/kiali-configmap.yaml]: Same "https" here.
I am not sure which one is the good one, at least in my installation of istio 1.0 I exposed the jaeger-query svc in non secure format.

File: handlers/grafana.go [hunk 6]
```diff
@@ -69,29 +69,13 @@ func getGrafanaInfo(osRouteSupplier osRouteSupplier, serviceSupplier serviceSupp
 		VarWorkload:           grafanaConfig.VarWorkload,
 	}
 
-	// Case of overridden URL from config: use it and return immediately
-	if grafanaInfo.URL != "" {
-		return &grafanaInfo, http.StatusOK, nil
+	if grafanaInfo.URL == "" {
```
Comment (lucasponce) [inline handlers/grafana.go]: Are these changes ok ?
I guess that we would like to check first if there is an URL defined via config, if not, then try to search it through the services API or openshift route API.
And I guess that Grafana and Jaeger services should use same mechanism.

**Comment by lucasponce** [comment]
@mwringe 
Please, put an eye on this PR too.
As discussed plan is to not infer grafana/jaeger external services as they can be published in several ways.
Main decision is to use the configuration and document this properly.

File: README.adoc (hunk 2 above)
Comment (aljesusg) [reply README.adoc]: Yes, I need to change it

**Comment by aljesusg** [comment]
comments addressed @lucasponce . Thanks

File: README.adoc [hunk 7]
```diff
@@ -460,26 +460,26 @@ external_services:
...
 ----
 external_services:
   jaeger:
-    service_namespace: VALUE
+    url: VALUE
 ----
 
-|`JAEGER_SERVICE`
-|The Kubernetes service name for Jaeger. (default is `jaeger-query`)
+|`JAEGER_SERVICE_NAMESPACE`
```
Comment (lucasponce) [inline README.adoc]: Do we need the JAEGER_SERVICE_NAMESPACE now ?

File: README.adoc [hunk 8]
```diff
@@ -460,26 +460,26 @@ external_services:
...
 ----
 external_services:
   jaeger:
-    service: VALUE
+    service_namespace: VALUE
 ----
 
-|`JAEGER_SERVICE_PORT`
-|The Kubernetes service port for Jaeger. (default is `16686`)
+|`JAEGER_SERVICE`
```
Comment (lucasponce) [inline README.adoc]: Same here, I don't think we need the JAEGER_SERVICE now.
Perhaps the info that "jaeger-query" is the service to expose should go into the JAEGER_URL variable.

File: README.adoc [hunk 9]
```diff
@@ -503,6 +503,73 @@ login_token:
 ----
 |===
 
+== Configure External Services
+
+=== Jaeger
+
+To configure the Jaeger Service you need to set the URL parameter in the config.yaml
```
Comment (lucasponce) [inline README.adoc]: I little more wording here, it should describe which service (jaeger-query) is necessary to expose.
So I would say, "The Jaeger Service (jaeger-service) should be exposed externally, and the URL should be set in the ..." (or something like that, forgive my English here).

File: README.adoc [hunk 10]
```diff
@@ -503,6 +503,73 @@ login_token:
...
+
+[source,yaml]
+----
+external_services:
+  jaeger:
+    url: https://jaeger-query-istio-system.127.0.0.1.nip.io
+----
+
+If you are using the Make task **openshift-deploy** or **k8s-deploy** you need to change in the **kiali-configmap.yaml** the value of the jaeger > url
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply README.adoc]: I miss a couple of commands after the configmap.yaml to tell users how to update this file and recreate kiali pods.
Users won't use the make task.
For example:
https://www.ibm.com/blogs/bluemix/2018/08/how-to-enable-kiali-for-istio-on-ibm-cloud-kubernetes-service/?social_post=1738063939&fst=Learn&linkId=55975565

**Comment by lucasponce** [comment]
Also, this needs a note in the getting started just to update the external services:
https://www.kiali.io/gettingstarted/
As it looks that "gettingstarted" section points to master/latest image and we should also comment how to point to grafana and jaeger.

**Comment by clyang82** [comment]
@aljesusg I saw you have successfully enabled SSL for jaeger ui in istio.

> url: https://jaeger-query-istio-system.127.0.0.1.nip.io

Could you share me how to achieve it? Thanks.

**Comment by lucasponce** [comment]
@clyang82 do you mean how to configure Kiali to use Jaeger or how to secure Jaeger itself ?
I think for Jaeger we have used the OpenShift Istio distribution that is responsible to package Jaeger.

I guess the SSL depends on the kubernetes distribution you are using.

Anyway I will cc @objectiser or @jpkrohling if they can add some additional info about it.

**Comment by jpkrohling** [comment]
If this is similar to our [main template](https://github.com/jaegertracing/jaeger-openshift/blob/master/production/jaeger-production-template.yml#L220), the OpenShift route is created with TLS by default.

**Comment by clyang82** [comment]
@lucasponce Thanks for your response. I mean how to secure Jaeger UI itself. I checked the Jaeger code and found that it seems Jaeger cannot support SSL itself.

**Comment by clyang82** [comment]
@jpkrohling Thanks for your suggestions. I am in Kubernetes environment so that there is no route.

**Pull request merged**